### DISCUSSION TITLE
Hyperlinks - Page.insert_link - how to provide a zoomed destination?

### DISCUSSION THREAD

**Discussion by AnweshGangula** [pymupdf/PyMuPDF#1251] [Q&A] [answered]
I would like to understand how does the zoom property work in [`Page.insert_link(linkdict)`](https://pymupdf.readthedocs.io/en/latest/page.html#Page.insert_link)

consider the linkdict = `{'kind': 1, 'xref': 121, 'from': Rect(112.31999969482422, 54.909912109375, 136.7917938232422, 68.35308837890625), 'page': 1, 'to': Point(108.0, 460.0), 'zoom': 0.0, 'id': 'VLRQFKMMYAYEKJIV'}`
what does the value 0.0 mean? 

- When I try to extract all links from a pdf ([Test Document - Github.pdf](https://github.com/pymupdf/PyMuPDF/files/7137961/Test.Document.-.Github.pdf)) which has 2 hyperlinks, all the zoom values are 0, but one of the links in the pdf navigates to a zoomed in potion of another page. 
- Is it possible to provide a `Rect` in the `to` object of the `dict` instead of a `Point`. This will remove the need of mentioning zoom value.
- i tried inserting a new link in each page with the following dict , but none of these are zoomed: 
`linkdict = {'kind': 1, 'xref': 121, 'from': Rect(112.31999969482422, 54.909912109375, 136.7917938232422, 68.35308837890625), 'page': 1, 'to': Point(108.0, 460.0), 'zoom': 4.0, 'id': 'VLRQFKMMYAYEKJIV'}`

I use Bluebeam Revu to edit my PDF and add hyperlinks.
I'm on Windows 10 19042.1052 build
PuMuPDF version: 1.18.14

**Comment by JorjMcKie** [comment]
>  what does the value 0.0 mean?

No magnification.

The value is used to define the following destination format (see page 582 in https://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/pdf_reference_1-7.pdf):
![grafik](https://user-images.githubusercontent.com/8290722/132727175-54f4bd44-3677-4e59-9b44-92d66ff553c6.png)

**Reply by AnweshGangula** [reply]
I have attached a sample pdf above. Using Page.getlinks() in that gives dictionaries with zoom always 0. But the pdf has a links with varying zoom levels.
And even using page.insert_link with zoom adds a hyperlink with doesn't have any zoom associated.

Can you help me understand the issue with that?

**Comment by JorjMcKie** [comment, marked as answer]
>  gives dictionaries with zoom always 0

Yes, this unfortunately is like so. MuPDF's support for link is limited in that respect. This goes even back to MuPDF v1.10. In this version, all information returned for internal links (GoTo) is the page number and the top-left point on the target - nothing else.

PyMuPDF's API is much richer and stems from pre v1.10 times. Since version v1.10 I am filling in values into the old API as best as is possible.
I will adjust the documentation to make this clear.

> page.insert_link with zoom adds a hyperlink with doesn't have any zoom associated

This is a bug. I forgot to accept a provided zoom value. Will be corrected in next version. The only supported format ever was `/XYZ` as indicated above.

As mentioned, `page.update_link()` in that respect is a "write-only" function: even once corrected, reading the link will nevertheless return zoom=0.

**Reply by AnweshGangula** [reply]
Thank you for the detailed explanation. Really appreciate it.
One last question I have is - Can you update PyMuPdf to accept a `Rect` in the `to` object of the dict instead of a `Point`?  This will remove the need of mentioning zoom value.

**Comment by JorjMcKie** [comment]
> Can you update PyMuPdf to accept a Rect in the to object of the dict instead of a Point? This will remove the need of mentioning zoom value.

But it would need two more coordinates for the rectangle. I guess I don't see what actually could be saved here.
It would also require additional plausi checks as to whether the rect fits into the target page. Which, for "GoToR", would definitely not be possible anyway.
And it would require to support whatever more of the half a dozen alternative PDF `/Fit...` specs for fitting the view to the target device - which I am not inclined to do. I chose the `/XYZ left top zoom` spec because it is the most general one.
And all that with the fact in mind, that this information is afterwards not extractable for being put into the dictionary.
Bottom line: I don't believe it would make sense.

**Reply by AnweshGangula** [reply]
the reason I was asking for rectangle it to able to specify the boundary of the zoom, which in return would mean you don't need to mention any zoom specifically. But as I understand from your comment, the `/XYZ left top zoom` is a general standard and it would definitely makes sense to follow that. 
Thank you for your time.
